### DISCUSSION TITLE
How can I cycle messages in a nested while loop?

### DISCUSSION THREAD

**Discussion by LewisHancock** [FLAMEGPU/FLAMEGPU2#971] [Q&A] [answered]
Hi, I am wanting to iterate through the same message list multiple times within a while-loop so that I may build a simple data network within the function. The rest of my code works fine (including the network building which I have temporally removed during bug fixing). However, I cannot get the nested loop to iterate through all the available messages in the message list given. I am not sure what is going on so any help on how this issue can be overcome would be greatly appreciated!

I am using **FLAME GPU 1** and have attached a report of the problem below. I'm dyslexic, so sorry for my horrible spelling... the spell checker of LibreOffice isn't working on my laptop.

[2022_11_11_nested_loop_question.pdf](https://github.com/FLAMEGPU/FLAMEGPU2/files/9990492/2022_11_11_nested_loop_question.pdf)

Many thanks,
Lewis

**Comment by mondus** [comment]
Hi Lewis. It is not possible to iterate message loops multiple times within the same agent function. To accomplish what you want to do there are a couple of options.

1) Use agent array variables to store data from messages as you iterate through them a single time. Your agent functions can iterate agent arrays as many times as you need.
2) Use multiple agent functions to iterate the message lists. You could use a process similar to agent sub models in FGPU2 to continue applying a function until some condition is met. The stable marriage and sugarscape models apply this principle with respect to movement.

Also you should try Flame GPU 2. Porting your model will most likely not be a huge difficulty. The software has much better software engineering and support.

**Reply by LewisHancock** [reply]
Thanks for the reply! Iterating the list once, storing locally and then running that through the while loop sounds like the best work-around for Flame GPU 1. In the meantime, I'll attempt to port my model into version 2 and see how that goes.

**Comment by ptheywood** [comment, marked as answer]
Within [FLAME GPU 1](https://github.com/FLAMEGPU/FLAMEGPU), message list iteration is subject to a number of restrictions:

+ All threads must participate in the iteration (i.e. `get_next_X_message` must be called the same number of times by all threads (and strictly speaking without branching between the calls)).
+ Message lists can only be iterated once (I believe, though multi-iteration may work for some message types).

Effectivley this means that you cannot have a `while(message)` loop within any other branching statement (`while`, or `if`), and it is just a limitation of FLAME GPU 1.

---

For your intended behaviour of building the network of agent connections there might be a few options to achieve this in FLAME GPU 1, but exactly how depends on a few thngs:

+ Does the network of connections change throughout your simulation? (i.e. does it need calculating once, or repeatedly?)
+ How large is the `binding_platform_update_priority` message list? (I assume its related to the population size?)

Depedning on exactly how this needs interacting with, you could:

+ Iterate the message list once, storing relevant information locally either in local variables within the agent function (which might severely harm performance) or within agent variable arrays (whcih will hurt performance, memory use and have a compile-time upper limit)
+ Use [Global Function Conditions](https://flamegpu.readthedocs.io/en/master/flamegpu/modelspec.html#global-function-conditions) to run agent functions N times, but this is not the cleanest to implement (We've revised this in FLAMEGPU 2 as submodels)
+ Compute it on the host (CPU) in serial, which will be slow but might be the only option depending on use-case.

FLAME GPU 2 does not have the same restrictions on iterating message lists together or only once, and has the majority of features from FLAME GPU 1 implemented, but will be a bit of effort to port to give n you have atleast 1400 lines of code.

**Reply by LewisHancock** [reply]
Thanks! I intend to re-build the network after every time step. However, since <<100 messages would typically be extracted from the binding_platform_update_priority message list by each agent, then I suppose iterating the messages once, storing locally and then build the network from this could be the easiest way.

I will also look into porting the model into Flame GPU 2 as it sounds like a good environment! My c code is ~2000 lines long in total so I agree it may take a bit of effort (although it's mainly the switch from an XML model file that I find the most intimidating)
